src/python/pants/engine/intrinsics.py

```diff
@@ -120,13 +121,39 @@ async def run_id() -> RunId:
     return await native_engine.run_id()
 
 
+# NB: Call one of the helpers below, instead of calling this rule directly,
+#  to ensure correct application of restartable logic.
 @_uncacheable_rule
-async def interactive_process(
+async def _interactive_process(
     process: InteractiveProcess, process_execution_environment: ProcessExecutionEnvironment
 ) -> InteractiveProcessResult:
     return await native_engine.interactive_process(process, process_execution_environment)
 
 
+async def run_interactive_process(iproc: InteractiveProcess) -> InteractiveProcessResult:
+    # NB: We must call mark_nonrestartable() in this helper, rather than in a nested @rule call,
+    #  so that the Task for the @rule that calls this helper is the one marked as non-restartable.
+    if not iproc.restartable:
+        mark_nonrestartable()
+
+    ret: InteractiveProcessResult = await _interactive_process(iproc)
+    return ret
+
+
+async def run_interactive_process_in_environment(
```

Comment:
Note that we need both helpers, as I  couldn't get a dynamic check for a None `EnvironmentName` to play nicely with the engine (and, to a lesser extent, with mypy). It caused the RuleVisitor and the solver to get confused and error.